pkg/automationconfig/automation_config.go

```diff
@@ -140,6 +141,31 @@ type Auth struct {
 	AutoPwd string `json:"autoPwd,omitempty"`
 }
 
+type CustomRole struct {
+	Role                       string                      `json:"role"`
+	DB                         string                      `json:"db"`
+	Privileges                 []Privilege                 `json:"privileges"`
+	Roles                      []Role                      `json:"roles"`
+	AuthenticationRestrictions []AuthenticationRestriction `json:"authenticationRestrictions,omitempty"`
+}
+
+type Privilege struct {
+	Resource Resource `json:"resource"`
+	Actions  []string `json:"actions"`
+}
+
+type Resource struct {
```

Comment:
The definition for a resource can be found here: https://docs.mongodb.com/manual/reference/resource-document/
It takes one of the following forms:
```
{db: string, collection: string}
{cluster: true}
{anyResource: true}
```

However, if either db or collection are set to "", that has a special meaning that indicates "all" databases/collections. Unfortunately in this case, the `omitempty` struct tag omits empty strings. If I mark db/collection as 'omitempty', the case where db and collection are set to "" gets sent to the agent as `{}`. Otherwise, if a user just tries to specify `{cluster: true}`, the agent will get `{db: "", collection: "", cluster: true}`, which is considered invalid.

Is there some way to omit these fields only when they are set to nil, for instance? That way we could preserve both the case where `{db: "", collection: ""}` and `{cluster: true}`.

Reply:
yes, we can achieve this using a string reference

```golang
Collection  *string `json:"collection,omitempty"`
```

omitempty will leave out the field if the underlying type has the zero value, and `""` is not the zero value for a pointer.

This is a pretty idiomatic way of dealing with optional fields when the normal zero value has a special meaning

Reply:
Thank you! Using a string reference everywhere instead worked like a charm. This is done now.

Reply:
Thank you! Using a string reference everywhere instead worked like a charm. This is done now.